Completely delegate handling of query args to the backend API

jblz opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

Instead of maintaining logic for manipulating arguments that will be sent along to the Parse.ly API, accept a single parameter called `query` and forward that along.

Additionally, this moves away from the `Recommended` or `Recommendations` nomenclature and settles on `related` instead.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

This is a move to simplify the API endpoint powering the block. The ultimate source of truth of what is acceptable is determined by the Parse.ly API service. If that is ever to change, this change will prevent us from also having to update the logic in this plugin to match.

As to the nomenclature, this aims for consistency between our symbols and the path of the remote API call.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

* Automated testing included in this PR
* Manual testing with #642 adapted to work against this new data shape format
* Manual testing in the browser, e.g. browse to: http://localhost:8888/wp-json/wp-parsely/v1/related?query[limit]=2&query[sort]=score

Note: for manual testing, you'll need to add the following in a separate plugin:
`add_filter( 'wp_parsely_enable_related_endpoint', '__return_true' );`

## Screenshots (if appropriate)

jblz committed:
Completeley delegate handling of query args to the backend API

GaryJones commented:
Right now, the class handles both the API call, and the caching of its own results.

I'd love to see this refactored to use a _Decorator pattern_ to provide the caching behaviour, to really separate out responsibilities.

jblz committed:
Merge branch 'develop' of github.com:Parsely/wp-parsely into add/recommendations-proxy-endpoint--proxy-all-query-args

jblz committed:
Use decorator pattern for cache proxy & rename everything to /related

jblz committed:
Merge branch 'develop' of github.com:Parsely/wp-parsely into add/recommendations-proxy-endpoint--proxy-all-query-args

jblz committed:
sort query args by key. reinstate test for get_api_url

jblz committed:
Add testing for caching decorator hit & miss

jblz committed:
use createMock instead

pauarge committed:
Merge branch 'develop' into add/recommendations-proxy-endpoint--proxy-all-query-args

pauarge commented on src/Endpoints/index.php:
```diff
@@ -0,0 +1,12 @@
+
+namespace Parsely;
+
+require_once __DIR__ . '/class-related-api-proxy.php';
```
Interesting approach for adding a loader pattern. I like it, but I'd rather do this for all files in the project rather than just this one. 

Given that's only one file for now, could we add it as usual (in `wp-parsely.php`) and then add this structure in a further PR?

pauarge commented on src/Endpoints/index.php:
```diff
@@ -0,0 +1,12 @@
+/**
+ * Loader for REST API endpoints served by this site
+ *
+ * @package Parsely
```
Could we add a `@since` annotation here?

pauarge commented on src/RemoteAPI/index.php:
```diff
@@ -0,0 +1,20 @@
+<?php
+/**
+ * Loader for Parsely Remote API functionality.
```
Could we add a `@since` annotation here?

jblz committed:
Merge branch 'develop' of github.com:Parsely/wp-parsely into add/recommendations-proxy-endpoint--proxy-all-query-args

jblz committed:
Move code sourcing to wp-parsely.php

jblz committed:
Add since annotation to rest_api_init hook

pauarge committed:
Merge branch 'develop' into add/recommendations-proxy-endpoint--proxy-all-query-args

GaryJones commented on src/Endpoints/class-related-api-proxy.php:
```diff
@@ -0,0 +1,121 @@
+ * Configure a REST API endpoint for use e.g. by the Related Block.
+ */
+final class Related_API_Proxy {
+	const OBJECT_CACHE_TTL = 5 * MINUTE_IN_SECONDS;
```
Now that there's a decorator, this constant seems to be unused.

GaryJones commented on src/Endpoints/class-related-api-proxy.php:
```diff
@@ -0,0 +1,121 @@
+							},
+						),
+					),
+				),
```
That's a lot of indentation. Suggest extract into variable, and use the variable in this call.

GaryJones commented on src/Endpoints/class-related-api-proxy.php:
```diff
@@ -0,0 +1,121 @@
+	/**
+	 * Determine if the endpoint can be called.
+	 *
+	 * @return boolean
```
```suggestion
	 * @return bool
```

GaryJones commented on src/Endpoints/class-related-api-proxy.php:
```diff
@@ -0,0 +1,121 @@
+	 * Cached "proxy" to the Parsely `/related` endpoint
+	 *
+	 * @param WP_Rest_Request $request The request object.
+	 * @return array
```
This method returns objects, not arrays.

GaryJones commented on src/RemoteAPI/related/class-related-caching-decorator.php:
```diff
@@ -0,0 +1,62 @@
+	 * @param Proxy           $proxy The Proxy object to cache.
+	 * @param WP_Object_Cache $cache The WordPress object cache instance.
+	 */
+	public function __construct( Proxy $proxy, WP_Object_Cache $cache ) {
```
The presence of the class and not an interface as the type declaration ties this directly to the `WP_Object_Cache` class, and any child classes of it. Another cache object couldn't be used, even though this class shouldn't care _where_ the data is cached.

The way around this is to create an adaptor interface and an adaptor which consumes an `WP_Object_Cache` object and exposes methods for the `->get()` and `->set()` methods which are used in this class.

(This is more about good OOP practices, and it can be added later if necessary, so not a blocker to this PR.)

Ref:
- https://devdocs.magento.com/guides/v2.4/extension-dev-guide/adapters.html
- https://designpatternsphp.readthedocs.io/en/latest/Structural/Adapter/README.html

GaryJones commented on src/RemoteAPI/related/class-related-caching-decorator.php:
```diff
@@ -0,0 +1,62 @@
+	 * Implements caching for the proxy interface.
+	 */
+	public function get_items() {
+		$items = $this->cache->get( $this->cache_key, 'wp-parsely' );
```
Can we make `wp-parsely` a class constant?

GaryJones commented on src/RemoteAPI/related/class-related-caching-decorator.php:
```diff
@@ -0,0 +1,62 @@
+		$items = $this->cache->get( $this->cache_key, 'wp-parsely' );
+		if ( ! is_object( $items ) ) {
+			$items = $this->decorated_proxy->get_items();
+			$this->cache->set( $this->cache_key, $items, 'wp-parsely', 5 * MINUTE_IN_SECONDS );
```
Can this duration argument be made a class constant as well, to better surface this value?

GaryJones commented on src/RemoteAPI/related/class-related-proxy.php:
```diff
@@ -0,0 +1,109 @@
+	 * Constructor.
+	 *
+	 * @param Parsely $parsely Parsely instance.
+	 * @param array   $query The query arguments to send to the remote API.
```
Can we at least define the shape of the array, even if we don't want to specify the keys? `array<string, mixed>` or something like that. (Likewise for the class property as well)

GaryJones commented on src/RemoteAPI/related/class-related-proxy.php:
```diff
@@ -0,0 +1,109 @@
+	 *
+	 * @since 3.2.0
+	 */
+	public function get_query() {
```
Missing a `@return` doc, and return type declaration.

GaryJones commented on src/RemoteAPI/related/class-related-proxy.php:
```diff
@@ -0,0 +1,109 @@
+	 * Implements the fetcher for the proxy interface.
+	 *
+	 * @since 3.2.0
+	 */
```
Missing a `@return` tag.

GaryJones commented on tests/Integration/RemoteAPI/RelatedRemoteAPITest.php:
```diff
@@ -0,0 +1,176 @@
+	 * @return array[]
+	 */
+	public function data_related_api_url(): array {
+		return array(
```
Consider using `yield` here instead of a giant array. See https://github.com/Parsely/wp-parsely/pull/383 which made the exact same switch.

GaryJones commented on src/RemoteAPI/related/class-related-caching-decorator.php:
```diff
@@ -0,0 +1,62 @@
+	public function __construct( Proxy $proxy, WP_Object_Cache $cache ) {
+		$this->decorated_proxy = $proxy;
+		$this->cache           = $cache;
+		$this->cache_key       = 'api_recos-' . wp_hash( wp_json_encode( $this->decorated_proxy->get_query() ) );
```
Does this cache key prefix need updating to reflect we're now using the nomenclature `related`?

GaryJones commented on src/RemoteAPI/related/class-related-proxy.php:
```diff
@@ -0,0 +1,109 @@
+	 *
+	 * @since 3.2.0
+	 */
+	public function get_query() {
```
I can't think of a better way to generate a unique cache key, so I think you're right that this needs to be exposed. As such, should it be added to the interface as well?

GaryJones replied:
I changed the approach, so this comment is no longer relevant.

GaryJones commented on src/RemoteAPI/related/class-related-proxy.php:
```diff
@@ -0,0 +1,109 @@
+		}
+
+		$body    = wp_remote_retrieve_body( $result );
+		$decoded = json_decode( $body );
```
All of the decoding, and checks, are technically a new task from the actual getting a response, so it could be handled with a `JSON_Decoder` class (which implements a `Decoder` interface, to allow for future expansion of retrieving XML, CSV, plain text, etc.). This would then centralise "the ability to decode some JSON and return an error if it fails" into one place, allowing multiple proxy classes to make use of it.

GaryJones replied:
Since we now have a single proxy point, then this line will only appear in one place - this could be switched out later on, but it's not a blocker to this PR.

GaryJones reviewed (changes requested):
I love the work to move to the caching decorator!

A few small and larger suggestions.

jblz committed:
move cache ttl constant to decorator

jblz committed:
pull get_items_args into an array

jblz committed:
fixup some phpdoc annotations

jblz committed:
use class const for cache group slug

jblz committed:
array shape in @param

jblz committed:
phpdoc fixes

jblz committed:
Merge branch 'add/recommendations-proxy-endpoint--proxy-all-query-args' of github.com:Parsely/wp-parsely into add/recommendations-proxy-endpoint--proxy-all-query-args

jblz committed:
Merge branch 'develop' of github.com:Parsely/wp-parsely into add/recommendations-proxy-endpoint--proxy-all-query-args

jblz committed:
array<string, mixed> for class prop

jblz committed:
rename cache key

jblz committed:
Use a WP_Object_Cache-compatible adapter

jblz committed:
convert data provider to iterable

GaryJones commented on src/RemoteAPI/related/class-related-caching-decorator.php:
```diff
@@ -42,10 +41,10 @@ class Related_Caching_Decorator implements Proxy {
+	 * @param Cache_Adapter $cache An object cache instance.
 	 */
-	public function __construct( Proxy $proxy, WP_Object_Cache $cache ) {
+	public function __construct( Proxy $proxy, Cache_Adapter $cache ) {
```
```suggestion
	public function __construct( Proxy $proxy, Cache $cache ) {
```

Cache_Adapter is the class, but the interface is what should be type declared here.

GaryJones replied:
The doc on this constructor, and `$cache` class property doc will also need updating.

GaryJones commented on src/RemoteAPI/related/class-related-caching-decorator.php:
```diff
@@ -42,10 +41,10 @@ class Related_Caching_Decorator implements Proxy {
 	 */
-	public function __construct( Proxy $proxy, WP_Object_Cache $cache ) {
+	public function __construct( Proxy $proxy, Cache_Adapter $cache ) {
 		$this->decorated_proxy = $proxy;
```
Since it's _this_ class that is doing the decorating, then this property could just be named as `$proxy`, rather than `$decorated_proxy`.

GaryJones reviewed (dismissed):
Rather than `Related_Caching_Decorator` as the name of the class that provides the caching behaviour for a proxy, I think `Cached_Proxy` would be a better nomenclature. That way, we're not leaking our choice of design pattern into the naming of the class (a _Decorator_ makes sense to us as engineers, but it has no standing as part of our business modelling here) and with a more general approach, then it could be easily used with other proxies (after all, we just want to add caching behaviour, regardless of the proxy being used):

```php
$cache = new Cache_Adaptor( $GLOBALS['wp_object_cache'] );
$related_proxy = new Cached_Proxy( new Related_Proxy( $this->parsely, $params['query'] ), $cache );
$another_proxy = new Cached_Proxy( new Another_Proxy( $this->parsely, $params['query'] ), $cache );
```

The only thing then would be to find a way to generate the unique keys. We could make the proxy interface have a `get_key()` method, and add that to the Proxy concrete classes. But this would mean the Proxy classes needing to know that it might be cached (it should have no knowledge of any decorations that might be applied to it).

So alternatively, have the Cached_Proxy do a hash of the object given to it, maybe extracting the prefix from the class name, so that it can generate a unique key that way, regardless of  whatever proxy was passed to it.

```php
class Cached_Proxy() {
...
	public function generate_key() {
		$prefix = strtolower( $this->proxy::class ); // Maybe limit to the first or last 10 characters?
		$unique = hash( $this->proxy );

		return $prefix . $unique;
	}
}
```

GaryJones committed:
Rename Adaptor class to be what we're adapting

While it's useful to know that a class is an adaptor for something, that can be shared in comments, while the class name reflects what the thing is that we're adapting to. This allows for other adaptors to potentially be created.

GaryJones committed:
Rename Decorator class to describe the behaviour

While it's useful to know that a class is a decorator for something, having the class name reflect the behaviour that's being added is more descriptive.

This also makes the `Cached_Proxy` be a decorator for all proxies, not just the Related Proxy.

GaryJones committed:
Tests: Minor tweaks

GaryJones committed:
Move query args to get_items() method parameter

We could have the same Related_Proxy class instantiated once, and pass the possibly-variable query args only at the time we need to get the items.

The caching generates the unique key at the time the get_items() is called, using the proxy object, and the query args.

GaryJones committed:
Remove instantiations from inside Related_API_Proxy

Due to previous changes, we can now inject a proxy object into the constructor of Related_API_Proxy, and don't have to instantiate the proxy, cache, and cached_proxy objects inside this Related_API_Proxy class. This means that the Related_API_Proxy class no longer has knowledge that a cache behaviour even exists for the proxy class, let alone where the result might be cached.

All the REST API class knows, is that it is given a proxy class, and it defers to that when items need to be retrieved.

The decision to use a cache, and what cache implementation to use, is now handled only in the composable root / main plugin file.

The integration test passes in a proxy object without the caching behaviour.

GaryJones committed:
CS fixes

GaryJones committed:
Miscellaneous fixes from manual testing

GaryJones commented:
As of [95a8aa2](https://github.com/Parsely/wp-parsely/pull/649/commits/95a8aa220f84faf723a7c7a207b43ffc6a8edb82), this is working for me locally, with various changes I've suggested now implemented.

Default endpoint: `http://localhost:10024/wp-json/wp-parsely/v1/related`
<img width="1325" alt="Screenshot 2022-02-12 at 16 12 21" src="https://user-images.githubusercontent.com/88371/153719833-184789fc-b94e-4efa-bbdd-b6be1ef2deed.png">

Endpoint with a query passed through: `http://localhost:10024/wp-json/wp-parsely/v1/related?query[limit]=2`
<img width="1325" alt="Screenshot 2022-02-12 at 16 18 58" src="https://user-images.githubusercontent.com/88371/153719837-983831be-20de-4407-bbee-3d6538995f9e.png">

I've not looked into the JSON_Decoder suggestion yet, but wanted to share the progress so far.

GaryJones committed:
Implement base proxy class

This moves the logic for creating a proxy to the Parse.ly API into an abstract base class, which only requires two protected class constants in a child class to enable new proxies.

GaryJones commented:
With [b79d11c](https://github.com/Parsely/wp-parsely/pull/649/commits/b79d11c39163ede462c72a488b25dd698dfa75bb), we now have the `/related` API being proxied with this class:

```php
class Related_Proxy extends Base_Proxy {
	protected const ENDPOINT     = 'https://api.parsely.com/v2/related';
	protected const QUERY_FILTER = 'wp_parsely_related_endpoint_args';
}
```

Other Parse.ly API endpoints that we want to proxy in the future, and pass all of the query args to, only need a new child class with two protected constants defined!

jblz commented on src/Endpoints/class-related-api-proxy.php:
```diff
@@ -0,0 +1,132 @@
+	 */
+	public function permission_callback(): bool {
+		// Unauthenticated.
+		return true;
```
Should we impose a nonce check or any rate limiting?  Any changes can be done in a follow up, but I thought I'd raise it here at least.

GaryJones replied:
Probably needs more thought, so would suggest opening an issue to follow-up.

GaryJones replied:
https://github.com/Parsely/wp-parsely/issues/667

jblz reviewed (commented):
Thanks so much for tidying this up @GaryJones. This really sets us on solid ground for adding additional endpoints.
